tests/e2e/browse-bills.spec.ts

```diff
@@ -0,0 +1,637 @@
+      const summaryContent = await page.textContent(
+        ".Summary__TitleFormat-sc-791f19-1"
+      )
+      if (summaryContent) {
+        expect(summaryContent.toLowerCase()).toContain(searchTerm.toLowerCase())
+      } else {
+        console.warn(
+          `Both full content and summary content for the first bill on page were null or did not contain the search term.`
+        )
+      }
+    }
+
+    await page.goBack()
+    await page.waitForTimeout(1000)
+  }
+
+  /**
+   * Function to click a random number of times on the "next page" button.
+   * @param page - The Playwright page object.
+   */
+  const clickNextPageRandomTimes = async (page: Page) => {
+    const randomClicks = Math.floor(Math.random() * 5) + 1
+    for (let i = 0; i < randomClicks; i++) {
+      const hasNextPage = await page.$(
+        "li.ais-Pagination-item--nextPage:not(.ais-Pagination-item--disabled)"
+      )
+      if (hasNextPage) {
+        await hasNextPage.click()
+        await page.waitForTimeout(1000)
+      } else {
+        break
+      }
+    }
+  }
+
+  test("find bills via text search", async ({ page }) => {
+    // Perform a search and wait for the results to change
+    const searchTerm = getRandomWord()
+
+    await performSearch(page, searchTerm)
+
+    const initialResultCount = await page.textContent(
+      ".ResultCount__ResultContainer-sc-3931e200-0"
+    )
+
+    await waitForResultsToChange(page, initialResultCount!)
+  })
+
+  test("search label test", async ({ page }) => {
+    // Perform a search and check that the category labels include the search term
+    const searchTerm = getRandomWord()
+
+    await performSearch(page, searchTerm)
+
+    const labelsTextContent = await getCategoryLabelsTextContent(page)
+
+    labelsTextContent.forEach((text, index) => {
+      console.log(`Label ${index + 1}: ${text}`)
+    })
+
+    const containsSearchTerm = labelsTextContent.some(textContent =>
+      textContent.toLowerCase().includes(searchTerm.toLowerCase())
+    )
+    expect(containsSearchTerm).toBe(true)
+  })
+
+  test("check the first bill on random page", async ({ page }) => {
+    // Perform a search and check the first bill on random pages multiple times
+    const searchTerm = getRandomWord()
+
+    await performSearch(page, searchTerm)
+
+    const initialResultCount = await page.textContent(
+      ".ResultCount__ResultContainer-sc-3931e200-0"
+    )
+
+    await waitForResultsToChange(page, initialResultCount!)
+
+    let times = 2
+    do {
+      await checkFirstBill(page, searchTerm)
+      await clickNextPageRandomTimes(page)
+      times--
+    } while (times > 0)
+  })
+
+  test("no results found", async ({ page }) => {
+    // Test to ensure the application handles no results found cases
+    const searchTerm = "nonexistentsearchterm12345"
+
+    await performSearch(page, searchTerm)
+
+    const noResultsMessage = await page.textContent(
+      ".NoResults__Container-sc-bf043801-0 .text-center"
```

Comment:
OOC, are these the locators that Playwright is generating via VS Code / whatever IDE integration you're using (e.g. https://playwright.dev/docs/getting-started-vscode#picking-a-locator)? More curious for the implications for other tests going forward than for this PR, but still curious. We should feel free to add `data-testid`s where necessary if it supports the testing effort.

Reply:
I retrieved the locators by inspecting the HTML code using Google Chrome's Developer Tools. The locators are based on the frontend HTML and CSS classes that are rendered by the browser. These can be generated by frontend frameworks like React and styled with CSS or CSS-in-JS libraries.

It will definitely make the test more reliable by adding data-testid attributes to key elements in our application to create more stable locators. This will help us ensure that our tests remain robust against UI changes.